cpp/src/arrow/util/rle_encoding_internal.h

```diff
@@ -299,385 +552,988 @@ class RleEncoder {
+    return {batch_size - null_count, null_count};
   }
 
-  ARROW_DCHECK_GE(bit_width_, 0);
-  int values_read = 0;
-  int values_remaining = batch_size - null_count;
+  constexpr BatchCounter(size_type values_count, size_type null_count) noexcept
+      : values_count_(values_count), null_count_(null_count) {}
 
-  // Assume no bits to start.
-  arrow::internal::BitRunReader bit_reader(valid_bits, valid_bits_offset,
-                                           /*length=*/batch_size);
-  arrow::internal::BitRun valid_run = bit_reader.NextRun();
-  while (values_read < batch_size) {
-    if (ARROW_PREDICT_FALSE(valid_run.length == 0)) {
-      valid_run = bit_reader.NextRun();
+  [[nodiscard]] constexpr size_type ValuesCount() const noexcept { return values_count_; }
+
+  [[nodiscard]] constexpr size_type ValuesRead() const noexcept { return values_read_; }
+
+  [[nodiscard]] constexpr size_type ValuesRemaining() const noexcept {
+    ARROW_DCHECK_LE(values_read_, values_count_);
+    return values_count_ - values_read_;
+  }
+
+  constexpr void AccrueReadValues(size_type to_read) noexcept {
+    ARROW_DCHECK_LE(to_read, ValuesRemaining());
+    values_read_ += to_read;
+  }
+
+  [[nodiscard]] constexpr size_type NullCount() const noexcept { return null_count_; }
+
+  [[nodiscard]] constexpr size_type NullRead() const noexcept { return null_read_; }
+
+  [[nodiscard]] constexpr size_type NullRemaining() const noexcept {
+    ARROW_DCHECK_LE(null_read_, null_count_);
+    return null_count_ - null_read_;
+  }
+
+  constexpr void AccrueReadNulls(size_type to_read) noexcept {
+    ARROW_DCHECK_LE(to_read, NullRemaining());
+    null_read_ += to_read;
+  }
+
+  [[nodiscard]] constexpr size_type TotalRemaining() const noexcept {
+    return ValuesRemaining() + NullRemaining();
+  }
+
+  [[nodiscard]] constexpr size_type TotalRead() const noexcept {
+    return values_read_ + null_read_;
+  }
+
+  [[nodiscard]] constexpr bool IsFullyNull() const noexcept {
+    return ValuesRemaining() == 0;
+  }
+
+  [[nodiscard]] constexpr bool IsDone() const noexcept { return TotalRemaining() == 0; }
+
+ private:
+  size_type values_count_ = 0;
+  size_type values_read_ = 0;
+  size_type null_count_ = 0;
+  size_type null_read_ = 0;
+};
+
+// The maximal unsigned size that a variable can fit.
+template <typename T>
+constexpr auto max_size_for_v =
+    static_cast<std::make_unsigned_t<T>>(std::numeric_limits<T>::max());
+
+/// Overload for GetSpaced for a single run in a RleDecoder
+template <typename Converter, typename BitRunReader, typename BitRun,
+          typename values_count_type, typename value_type>
+auto RunGetSpaced(Converter& converter, typename Converter::out_type* out,
+                  values_count_type batch_size, values_count_type null_count,
+                  BitRunReader&& validity_reader, BitRun&& validity_run,
+                  RleDecoder<value_type>& decoder)
+    -> std::pair<values_count_type, values_count_type> {
```

Comment:
An opaque `std::pair` isn't very nice for the reader, especially when both members have the same type, perhaps make it a simple `struct`?